File: behavioral-design-pattern/chain-of-responsibility-pattern/chain_of_responsibility.ts
/**
 * Chain of Responsibility Pattern
 * ================================
 * Category: Behavioral Design Pattern
 *
 * Intent:
 *   Pass a request along a chain of handlers. Each handler either processes
 *   the request or forwards it to the next handler in the chain.
 *
 * When to use:
 *   - HTTP middleware pipelines, event processing, support ticket escalation
 *   - When multiple objects can handle a request and the handler isn't known in advance
 *   - When handlers should be configurable dynamically
 *
 * Key Participants:
 *   - Handler (abstract): Declares handle method and next reference
 *   - ConcreteHandler: Handles specific requests, forwards the rest
 *   - Client: Sends request to the first handler
 */

// ---------------------------------------------------------------------------
// Request data
// ---------------------------------------------------------------------------
enum Priority {
  LOW = 1,
  MEDIUM = 2,
  HIGH = 3,
  CRITICAL = 4,
}

interface SupportTicket {
  id: string;
  title: string;
  priority: Priority;
  description: string;
  handledBy: string | null;
}

// ---------------------------------------------------------------------------
// Handler base class
// ---------------------------------------------------------------------------
abstract class SupportHandler {
  private next: SupportHandler | null = null;

  setNext(handler: SupportHandler): SupportHandler {
    this.next = handler;
    return handler; // Enables chaining: a.setNext(b).setNext(c)
  }

  handle(ticket: SupportTicket): SupportTicket {
    if (this.canHandle(ticket)) {
      return this.process(ticket);
    }
    if (this.next) {
      return this.next.handle(ticket);
    }
    console.log(`  [Unhandled] Ticket ${ticket.id} reached end of chain`);
    return ticket;
  }

  protected abstract canHandle(ticket: SupportTicket): boolean;
  protected abstract process(ticket: SupportTicket): SupportTicket;
}

// ---------------------------------------------------------------------------
// Concrete Handlers
// ---------------------------------------------------------------------------
class AutoResponder extends SupportHandler {
  protected canHandle(ticket: SupportTicket): boolean {
    return ticket.priority === Priority.LOW;
  }
  protected process(ticket: SupportTicket): SupportTicket {
    ticket.handledBy = "AutoResponder (Bot)";
    console.log(`  [AutoResponder] Ticket ${ticket.id}: Sent automated FAQ response`);
    return ticket;
  }
}

class Level1Support extends SupportHandler {
  protected canHandle(ticket: SupportTicket): boolean {
    return ticket.priority === Priority.MEDIUM;
  }
  protected process(ticket: SupportTicket): SupportTicket {
    ticket.handledBy = "Level 1 Support Agent";
    console.log(`  [Level1] Ticket ${ticket.id}: Assigned to support agent`);
    return ticket;
  }
}

class Level2Support extends SupportHandler {
  protected canHandle(ticket: SupportTicket): boolean {
    return ticket.priority === Priority.HIGH;
  }
  protected process(ticket: SupportTicket): SupportTicket {
    ticket.handledBy = "Level 2 Senior Engineer";
    console.log(`  [Level2] Ticket ${ticket.id}: Escalated to senior engineer`);
    return ticket;
  }
}

class ManagerSupport extends SupportHandler {
  protected canHandle(ticket: SupportTicket): boolean {
    return ticket.priority === Priority.CRITICAL;
  }
  protected process(ticket: SupportTicket): SupportTicket {
    ticket.handledBy = "Support Manager";
    console.log(`  [Manager] Ticket ${ticket.id}: CRITICAL — Manager notified`);
    return ticket;
  }
}

// ---------------------------------------------------------------------------
// Usage Example
// ---------------------------------------------------------------------------
const auto = new AutoResponder();
const l1 = new Level1Support();
const l2 = new Level2Support();
const manager = new ManagerSupport();

// Build the chain
auto.setNext(l1).setNext(l2).setNext(manager);

const tickets: SupportTicket[] = [
  { id: "T-001", title: "Password reset", priority: Priority.LOW, description: "Can't remember", handledBy: null },
  { id: "T-002", title: "App crashes", priority: Priority.MEDIUM, description: "Error on iOS", handledBy: null },
  { id: "T-003", title: "Data corruption", priority: Priority.HIGH, description: "Inconsistent", handledBy: null },
  { id: "T-004", title: "Service outage", priority: Priority.CRITICAL, description: "Prod down!", handledBy: null },
];

console.log("Processing support tickets through the chain:\n");
for (const ticket of tickets) {
  const result = auto.handle(ticket);
  console.log(`    -> Handled by: ${result.handledBy}\n`);
}

File: behavioral-design-pattern/chain-of-responsibility-pattern/chain_of_responsibility.py
"""
Chain of Responsibility Pattern
================================
Category: Behavioral Design Pattern

Intent:
    Avoid coupling the sender of a request to its receiver by giving multiple
    objects a chance to handle the request. Chain the receiving objects and pass
    the request along the chain until an object handles it.

When to use:
    - When multiple objects can handle a request and the handler isn't known in advance
    - HTTP middleware pipelines, event processing, support ticket escalation
    - When you want to decouple request senders from receivers
    - When the set of handlers should be configurable dynamically

Key Participants:
    - Handler (interface): Declares method for handling and a reference to the next handler
    - ConcreteHandler: Handles requests it's responsible for, forwards the rest
    - Client: Initiates the request to the first handler in the chain
"""

from abc import ABC, abstractmethod
from dataclasses import dataclass
from enum import IntEnum


# ---------------------------------------------------------------------------
# Request data
# ---------------------------------------------------------------------------
class Priority(IntEnum):
    LOW = 1
    MEDIUM = 2
    HIGH = 3
    CRITICAL = 4


@dataclass
class SupportTicket:
    id: str
    title: str
    priority: Priority
    description: str
    handled_by: str | None = None


# ---------------------------------------------------------------------------
# Handler interface
# ---------------------------------------------------------------------------
class SupportHandler(ABC):
    """
    Each handler in the chain either handles the ticket or passes it
    to the next handler. This is the core of the pattern.
    """

    def __init__(self) -> None:
        self._next: SupportHandler | None = None

    def set_next(self, handler: "SupportHandler") -> "SupportHandler":
        """Chain handlers together. Returns the next handler for fluent chaining."""
        self._next = handler
        return handler

    def handle(self, ticket: SupportTicket) -> SupportTicket:
        """
        Try to handle the ticket. If this handler can't (or shouldn't),
        pass it to the next handler in the chain.
        """
        if self.can_handle(ticket):
            return self.process(ticket)
        elif self._next:
            return self._next.handle(ticket)
        else:
            # End of chain — no one handled it
            print(f"  [Unhandled] Ticket {ticket.id} reached end of chain")
            return ticket

    @abstractmethod
    def can_handle(self, ticket: SupportTicket) -> bool:
        ...

    @abstractmethod
    def process(self, ticket: SupportTicket) -> SupportTicket:
        ...


# ---------------------------------------------------------------------------
# Concrete Handlers — support escalation chain
# ---------------------------------------------------------------------------
class AutoResponder(SupportHandler):
    """Handles LOW priority tickets with automated responses."""

    def can_handle(self, ticket: SupportTicket) -> bool:
        return ticket.priority == Priority.LOW

    def process(self, ticket: SupportTicket) -> SupportTicket:
        ticket.handled_by = "AutoResponder (Bot)"
        print(f"  [AutoResponder] Ticket {ticket.id}: Sent automated FAQ response")
        return ticket


class Level1Support(SupportHandler):
    """Handles MEDIUM priority tickets — basic troubleshooting."""

    def can_handle(self, ticket: SupportTicket) -> bool:
        return ticket.priority == Priority.MEDIUM

    def process(self, ticket: SupportTicket) -> SupportTicket:
        ticket.handled_by = "Level 1 Support Agent"
        print(f"  [Level1] Ticket {ticket.id}: Assigned to support agent for troubleshooting")
        return ticket


class Level2Support(SupportHandler):
    """Handles HIGH priority tickets — specialized technical support."""

    def can_handle(self, ticket: SupportTicket) -> bool:
        return ticket.priority == Priority.HIGH

    def process(self, ticket: SupportTicket) -> SupportTicket:
        ticket.handled_by = "Level 2 Senior Engineer"
        print(f"  [Level2] Ticket {ticket.id}: Escalated to senior engineer")
        return ticket


class ManagerSupport(SupportHandler):
    """Handles CRITICAL tickets — manager intervention required."""

    def can_handle(self, ticket: SupportTicket) -> bool:
        return ticket.priority == Priority.CRITICAL

    def process(self, ticket: SupportTicket) -> SupportTicket:
        ticket.handled_by = "Support Manager"
        print(f"  [Manager] Ticket {ticket.id}: CRITICAL — Manager notified immediately")
        return ticket


# ---------------------------------------------------------------------------
# Usage Example
# ---------------------------------------------------------------------------
if __name__ == "__main__":
    # Build the chain: AutoResponder → Level1 → Level2 → Manager
    auto = AutoResponder()
    l1 = Level1Support()
    l2 = Level2Support()
    manager = ManagerSupport()

    # Chain them together (fluent API)
    auto.set_next(l1)
    l1.set_next(l2)
    l2.set_next(manager)

    # Create test tickets
    tickets = [
        SupportTicket("T-001", "Password reset", Priority.LOW, "Can't remember password"),
        SupportTicket("T-002", "App crashes on login", Priority.MEDIUM, "Error on iOS 17"),
        SupportTicket("T-003", "Data corruption", Priority.HIGH, "User data inconsistent"),
        SupportTicket("T-004", "Full service outage", Priority.CRITICAL, "Production down!"),
    ]

    # Send each ticket through the chain — each finds its appropriate handler
    print("Processing support tickets through the chain:\n")
    for ticket in tickets:
        result = auto.handle(ticket)  # Always start at the first handler
        print(f"    → Handled by: {result.handled_by}\n")
